### PR TITLE
Lazy create StaticFileMiddleware statemachine

### PR DESCRIPTION
**Pull request by benaadams**
* Lazy create StaticFileMiddleware statemachine

@rynowak @JamesNK @Tratcher question...

`dotnet new mvc` etc adds `app.UseStaticFiles();` prior to `app.UseRouting()` and `app.UseEndpoints()`  (also before auth) is this desirable?

https://github.com/aspnet/AspNetCore/blob/8b536b08ca76126e58f662b7f500bb191116f2bd/src/ProjectTemplates/Web.ProjectTemplates/content/StarterWeb-CSharp/Startup.cs#L157-L174

e.g. it outputs
```csharp
app.UseStaticFiles();

app.UseCookiePolicy();

app.UseRouting();

app.UseAuthorization();

app.UseEndpoints(endpoints =>
{
    endpoints.MapControllerRoute(
        name: "default",
        pattern: "{controller=Home}/{action=Index}/{id?}");
    endpoints.MapRazorPages();
});
```
For the call chain

image

### CODE REVIEW

**Comment by rynowak** [comment]
> dotnet new mvc etc adds app.UseStaticFiles(); prior to app.UseRouting() and app.UseEndpoints() (also before auth) is this desirable?

Yes. Static files short circuits when it matches and doesn't interact with any of these things. So it makes sense to put before.

**Review by rynowak — APPROVED** [review]
From me, but @Tratcher should look as well.

**Comment by benaadams** [comment]
@Tratcher the ignore whitespace diff is clearest to look at

**Review by Tratcher — APPROVED** [review]

File: src/Middleware/StaticFiles/src/StaticFileMiddleware.cs [hunk 1]
```diff
@@ -95,52 +95,57 @@ public async Task Invoke(HttpContext context)
...
-                        await fileContext.SendStatusAsync(Constants.Status412PreconditionFailed);
+                        await fileContext.SendAsync();
+                        _logger.FileServed(fileContext.SubPath, fileContext.PhysicalPath);
                         return;
+                    }
+                    catch (FileNotFoundException)
+                    {
+                        context.Response.Clear();
+                    }
+                    break;
```
Comment (Tratcher) [inline src/Middleware/StaticFiles/src/StaticFileMiddleware.cs]: This case used to fall though to next but now returns OK. It's causing test failures.

**Review by Tratcher — CHANGES_REQUESTED** [review]

File: src/Middleware/StaticFiles/src/StaticFileMiddleware.cs (hunk 1 above)
Comment (benaadams) [reply src/Middleware/StaticFiles/src/StaticFileMiddleware.cs]: Ah, good spot, added

**Review by Tratcher — APPROVED** [review]

**Pull request merged**